docs/50_writing_tests.md

```diff
@@ -0,0 +1,463 @@
+
+The TestData class can be made to cover this by overriding the configureExclusions() method to add instances of the TestData classes for the exclusion devices to the object's internal list of exclusions.
+
+Detection code:
+
+```
+} else if (b.name().startsWith(QStringLiteral("ECH")) && !echelonRower && !echelonStride &&
+                       !echelonConnectSport && filter) {
+```
+The configureExclusions code is overridden to specify the exclusion test data objects. Note that the test for a previously detected device of the same type is not included.
+
+```
+#pragma once
+
+#include "Devices/Bike/biketestdata.h"
+#include "Devices/EchelonRower/echelonrowertestdata.h"
+#include "Devices/EchelonStrideTreadmill/echelonstridetreadmilltestdata.h"
+#include "devices/echelonconnectsport/echelonconnectsport.h"
+
+class EchelonConnectSportBikeTestData : public BikeTestData {
+
+public:
+    EchelonConnectSportBikeTestData() : BikeTestData("Echelon Connect Sport Bike") {
+        this->addDeviceName("ECH", comparison::StartsWith);
+    }
+
+    void configureExclusions() override {
+        this->exclude(new EchelonRowerTestData());
+        this->exclude(new EchelonStrideTreadmillTestData());
+    }
+
+    deviceType get_expectedDeviceType() const override { return deviceType::EchelonConnectSport; }
+
+    bool get_isExpectedDevice(bluetoothdevice * detectedDevice) const override {
+        return dynamic_cast<echelonconnectsport*>(detectedDevice)!=nullptr;
+    }
+};
+
+
+```
+
+### When a single TestData Class Can't Cover all the Conditions
+
+Detection code:
+
+```
+QString powerSensorName =
+        settings.value(QZSettings::power_sensor_name, QZSettings::default_power_sensor_name).toString();
+...
+} else if ((b.name().toUpper().startsWith(QStringLiteral("STAGES ")) ||
+                        (b.name().toUpper().startsWith("TACX SATORI")) ||
+                        ((b.name().toUpper().startsWith("KICKR CORE")) && !deviceHasService(b, QBluetoothUuid((quint16)0x1826)) && deviceHasService(b, QBluetoothUuid((quint16)0x1818))) ||
+                        (b.name().toUpper()==QStringLiteral("QD")) ||
+                        (b.name().toUpper().startsWith(QStringLiteral("ASSIOMA")) &&
+                         powerSensorName.startsWith(QStringLiteral("Disabled")))) &&
+                       !stagesBike && !ftmsBike && filter) {
+```
+
+This presents 3 scenarios for the current test framework. 
+1. Match names only (starts with:"STAGES ", starts with: "TACX SATORI", equals: "QD")
+2. Match the name "KICKR CORE", presence and absence of specific service ids
+3. Match the name "ASSIOMA" and the power sensor name setting starts with "Disabled"
+					   
+The framework is not currently capable of specifying all these scenarios in a single class. 
+The generated test data is approximately the combinations of these lists: names * settings * bluetoothdeviceInfo * exclusions.
+If a combination should not exist, a separate class should be used.
```

Comment:
@cagnulein I think this test framework would be easier to understand if instead of "here's a device class, here are the configs that should result in its detection, check that they do; here are configs that shouldn't, check that they don't" ;  "here's a set of configurations corresponding to a real world device, and an expected device class, check that these configurations cause the device class to be created; invalidate the configurations and check they don't". 

I think this would be longer, but easier to understand.

Reply:
I'm not looking to do this anytime soon.